### ISSUE TITLE
Kubelet: We should separate status update with SyncPod

### ISSUE THREAD

**Issue by Random-Liu** [kubernetes/kubernetes#25231]
As is discussed in https://github.com/kubernetes/kubernetes/issues/19077#issuecomment-216686379.

Now we only update pod status at the beginning of `SyncPod`, so the pod status could never be able to reflect the changes happening in `SyncPod`, for example image pulling progress.

We should separate status update with `SyncPod`, so that we could update pod status whenever we need, and possibly driven by events.

/cc @kubernetes/sig-node

**Comment by resouer** [comment]
Isn't this can be fix by moving `kl.statusManager.SetPodStatus()` to a separate function? Or I missed something ...

**Comment by yujuhong** [comment]
> Isn't this can be fix by moving kl.statusManager.SetPodStatus() to a separate function? Or I missed something ...

Yes, with some caveats. The function will be triggered by events. In which case, kubelet may send out more status updates than it does now. We have to measure and see if we need to batch them.

**Comment by resouer** [comment]
Sorry, @yujuhong , I did not dig this deeply but I can't tell the difference between call `SetPodStatus` directly when needed VS send out a event to call `SetPodStatus` when needed., it seems `SetPodStatus` is expected to work in a sync way. 

Or, what we want is replace the current  "update pod status at will " way with a unified process like generating a new pleg event during `relist` (though I'm not sure if `relist` is enough to cover all pod status update cases honestly) ...

**Comment by yujuhong** [comment]
In the current implementation, the pod status will be updated whenever a pod worker syncs the pod. A pod worker can be woken up by a variety of things, e.g., pleg events, 1 min resync period, etc.
This issue is to decouple "setting the pod status" with "pod worker syncing the pod". This means that the other events can cause kubelet to set the pod status, e.g., to periodically report image pulling progress. Of course, we still need to update the status whenever there are any pleg events as well.

**Comment by alejandrox1** [comment]
From my understanding, this needs to be done in order to tackle https://github.com/kubernetes/kubernetes/issues/19077
/reopen
/lifecycle frozen

**Comment by tedyu** [comment]
Should status update be triggered when either 

1. there is new event and predefined interval has passed
2. SyncPod is called

#1 is for batching and #2 is to be compatible with current implementation.

**Comment by markusthoemmes** [comment]
To add to this: It'd be super useful if we could surface the pod's IP in the status as soon as we actually know it (rather than waiting for a PLEG relist) as that'd enable endpoints and services to start picking up on it while the kubelet is doing its job to complete the pod's readiness.

I'd be very happy to pick this up if anybody would help guide me a little in the process.
